# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 1304 Мусаев Артур. Вариант №42.

## Цель работы

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения

### Понятие аддитивной цепочки

Аддитивной цепочкой натурального числа $n$ называется последовательность натуральных чисел<br> $$ 1=a_0,a_1,a_2,\dots,a_m=n,\quad   a_i=a_j+a_k, \quad \forall k \leq j < i$$ Число $r$ называется длиной цепочки.<br>Длина кратчайшей цепочки для заданного $n$ обозначается $l(n) = r$<br>
Цепочка называется звёздной, если каждый её элемент получается таким образом:<br>
$$a_i=a_{i-1}+a_k\quad \forall k<i$$
Справедливы некоторые оценки для $l(n)$:<br>
$$l(n) \leq \lambda(n) +\nu(n) - 1$$
$$l(n) \geq \lceil \log_{2}(n) \rceil$$
    
### Алгоритм Яо

Алгоритм Яо позволяет вычислить аддитивную цепочку $n\in \mathbb{N}$.<br>
Берётся произвольная $k > 1$, а $n$ раскладывается в $2^{k}$-ой СС:  
$$n = \sum_{i = 0}^{j}a_i2^{ik}, \quad a_j \neq 0$$

Введём функцию $$d(z) = \sum_{i:a_i = z}2^{ik}$$
Представим шаги алгоритма:
<ol>
    <li>Составим базовую последовательность $\{1, 2, 4, \dots, 2^{\lambda(n)}\}$</li>
    <li>Вычисляем значения $d(z)$ для всех $z \in \{1, 2, 4, \dots,2^{k-1}\}, \quad d(z) \neq 0$</li>
    <li>Вычисляем $zd(z) \quad   \forall z$</li>
    <li>В итоге получится разложение $n = \sum_{z=1}^{2^k-1}zd(z)$</li>
    <li>Добавим необходимые шаги из разложения в пункте 3 в базовую последовательность</li>
</ol>
Алгоритм не гарантирует минимальности аддитивной цепочки.<br>


### Алгоритм дробления вектора индексов
С помощью данного алгоритма можно найти минимальную звёздную цепочку для числа $n$. Введём некоторые понятия, укажем шаги работы алгоритма.<br>
Вектором индексов называется последовательность $r_1,r_2,\dots,r_{m-1}$, где $$r_i=\{z:1\leq z\leq i\}, \quad a_i=a_{i-1}+a_{r_{i-1}},\quad2\leq i\leq m-1.$$
Зададим два вектора индексов $r=\{r_i\}_{i=1}^{m-1}$ и $\widetilde r=\{\widetilde r_i\}_{i=1}^{m-1}$. Тогда $r \succ \widetilde r$, если $r_1=\widetilde r_1,r_2=\widetilde r_2,\dots,r_{m-1}=\widetilde r_{m-1}$, а $r_m>\widetilde r_m$.<br>
Рассмотрим вектор индекса вида $\{r_1, r_2, \dots, r_q\} \cup \{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}.$ Назовём левую часть фиксированной, а правую - меняющейся.<br>
Значение последнего элемента цепочки наибольшее при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{q+1,\dots,m\}\quad a_{max}=a_{q+1}*2^{m-q}$ <br>
Значение последнего элемента цепочки наименьшее при векторе индексов $\{r_1, r_2, \dots, r_q\}\cup\{1,\dots,1\}\quad a_{min}=a_{q+1}+m-q$  
Рассмотрим алгоритм дробления вектора индексов:
<ol>
    <li>Запускаем внешний цикл по длинам $\underline l(n) \le m \le \overline l(n)$, выбираем индекс дроблений $q=\frac{m}{2}$</li>
    <li>Внутренний цикл - перебор всех фиксированных частей. На каждом шагу вычисляем $a_{min}, a_{max}$ и строим цепочку.
        <ol>
          <li>Если $a_{m} = n$, то задача решена.</li>
          <li>Если $n \notin [a_{min}, a_{max}]$, то переходим к следующему набору $\{r_1, r_2, \dots, r_q\}$</li>
          <li>Если $n \in [a_{min}, a_{max}]$, то организуем внутренний цикл перебора меняющейся части.
                    <ol>
          <li>Если $a_{m} = a_{n}$, то задача решена</li>
          <li>Иначе переходим к следующему (по введённой упорядоченности) фиксированной части $\{ \rho_{q+1}, \rho_{q+2},\dots, \rho_{m}\}$</li>
         </ol>
            </li>
        </ol>
    </li>
    <li>Если все наборы фиксированной длины исчерпаны, то увеличиваем их длину во внешнем цикле</li>
</ol>

### Гипотеза Шольца-Брауэра
Гипотеза заключается в следующем:
$$l(2^{n}-1) \le l(n) + n - 1$$
При этом:
<ul>
    <li>Гипотеза доказана для звёздных цепочек</li>
    <li>Гипотеза справедлива для всех $n < 578469$</li>
    <li>Равенство выполняется для всех $n <= 64$</li>
</ul>

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной математики SageMath, провести анализ алгоритмов. Полученные результаты
содержательно проинтерпретировать.

## Выполнение работы

### Алгоритм Яо

Реализуем алгоритм Яо с соответствием теоритическим положениям:

In [80]:
def second_part_chain(z, dz):
    summ = dz
    res = [dz]
    bits = bin(z)[3:]
    for bit in bits:
        summ += summ
        res.append(summ)
        if bit == '1':
            summ += dz
            res.append(summ)
    return res

def d(Y, mods, z, k):
    res = 0
    for i in range(len(mods)):
        if mods[i] == z:
            res += 2**(i * k)
            if res not in Y:
                Y.append(res)
    return res

def yao(n, k):
    Y = [2**i for i in range(len(bin(n)[2:]))]
    mods = []
    koef = 2**k
    summ_zdz = 0
    n_ = n
    while n_ != 0:
        mods.append(n_ % koef)
        n_ = n_ // koef
    for z in range(1, 2**k):
        dz = d(Y, mods, z, k)
        if dz != 0:
            adds = second_part_chain(z, dz)
            for elem in adds:
                if elem not in Y:
                    Y.append(elem)
            summ_zdz = summ_zdz + adds[-1]
            if summ_zdz not in Y:
                Y.append(summ_zdz)
    return Y

Ниже реализован вывод длины аддитивной цепочки, полученной алгоритмом Яо и минимальной аддитивной цепочки для заданных n и k.

Продемонстрируем работу алгоритма на $n = 23, 54, 197, 384, 416, 675, 694, 887$ и $k=2,3,4,5,6$, при этом будем варьировать параметр $k$, сравнивать длину получившейся цепочки с длиной минимальной цепочки.

In [83]:
def test_yao():
    k_list = [2, 3, 4, 5, 6]
    n_list = {
              23:[1, 2, 4, 5, 9, 18, 23], 
              54:[1, 2, 4, 8, 16, 18, 36, 54], 
              197:[1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197], 
              384:[1, 2, 4, 8, 16, 32, 64, 128, 256, 384], 
              416:[1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416], 
              675:[1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675], 
              694:[1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694], 
              887:[1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887]
             }
    for n in n_list:
        for k in k_list:
            yao_result = yao(n, k)
            length = len(yao_result)
            #print(f"{k}|{n}|{yao_result}|{n_list[n]}|{length}|{len(n_list[n])}")
test_yao()

Оформим в виде таблицы:

| k |  n  |                                           Цепочка Яо                                           |                    Минимальная цепочка                   | Длина Яо | Длина минимальной |
|:-:|:---:|:----------------------------------------------------------------------------------------------:|:--------------------------------------------------------:|:--------:|:-----------------:|
| 2 |  23 |                                   [1, 2, 4, 8, 16, 20, 3, 23]                                  |                  [1, 2, 4, 5, 9, 18, 23]                 |     8    |         7         |
| 3 |  23 |                                  [1, 2, 4, 8, 16, 3, 6, 7, 23]                                 |                  [1, 2, 4, 5, 9, 18, 23]                 |     9    |         7         |
| 4 |  23 |                                  [1, 2, 4, 8, 16, 3, 6, 7, 23]                                 |                  [1, 2, 4, 5, 9, 18, 23]                 |     9    |         7         |
| 5 |  23 |                               [1, 2, 4, 8, 16, 5, 10, 11, 22, 23]                              |                  [1, 2, 4, 5, 9, 18, 23]                 |    10    |         7         |
| 6 |  23 |                               [1, 2, 4, 8, 16, 5, 10, 11, 22, 23]                              |                  [1, 2, 4, 5, 9, 18, 23]                 |    10    |         7         |
| 2 |  54 |                                 [1, 2, 4, 8, 16, 32, 6, 48, 54]                                |               [1, 2, 4, 8, 16, 18, 36, 54]               |     9    |         8         |
| 3 |  54 |                               [1, 2, 4, 8, 16, 32, 9, 18, 27, 54]                              |               [1, 2, 4, 8, 16, 18, 36, 54]               |    10    |         8         |
| 4 |  54 |                               [1, 2, 4, 8, 16, 32, 48, 3, 6, 54]                               |               [1, 2, 4, 8, 16, 18, 36, 54]               |    10    |         8         |
| 5 |  54 |                             [1, 2, 4, 8, 16, 32, 5, 10, 11, 22, 54]                            |               [1, 2, 4, 8, 16, 18, 36, 54]               |    11    |         8         |
| 6 |  54 |                         [1, 2, 4, 8, 16, 32, 3, 6, 12, 13, 26, 27, 54]                         |               [1, 2, 4, 8, 16, 18, 36, 54]               |    13    |         8         |
| 2 | 197 |                           [1, 2, 4, 8, 16, 32, 64, 128, 5, 192, 197]                           |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197]       |    11    |         11        |
| 3 | 197 |                           [1, 2, 4, 8, 16, 32, 64, 128, 192, 5, 197]                           |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197]       |    11    |         11        |
| 4 | 197 |                       [1, 2, 4, 8, 16, 32, 64, 128, 5, 48, 96, 192, 197]                       |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197]       |    13    |         11        |
| 5 | 197 |                         [1, 2, 4, 8, 16, 32, 64, 128, 5, 96, 192, 197]                         |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197]       |    12    |         11        |
| 6 | 197 |                           [1, 2, 4, 8, 16, 32, 64, 128, 192, 5, 197]                           |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 196, 197]       |    11    |         11        |
| 2 | 384 |                             [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]                            |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]         |    10    |         10        |
| 3 | 384 |                          [1, 2, 4, 8, 16, 32, 64, 128, 256, 192, 384]                          |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]         |    11    |         10        |
| 4 | 384 |                             [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]                            |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]         |    10    |         10        |
| 5 | 384 |                        [1, 2, 4, 8, 16, 32, 64, 128, 256, 96, 192, 384]                        |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]         |    12    |         10        |
| 6 | 384 |                          [1, 2, 4, 8, 16, 32, 64, 128, 256, 192, 384]                          |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 384]         |    11    |         10        |
| 2 | 416 |                        [1, 2, 4, 8, 16, 32, 64, 128, 256, 80, 160, 416]                        |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416]       |    12    |         11        |
| 3 | 416 |                        [1, 2, 4, 8, 16, 32, 64, 128, 256, 192, 384, 416]                       |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416]       |    12    |         11        |
| 4 | 416 |                        [1, 2, 4, 8, 16, 32, 64, 128, 256, 80, 160, 416]                        |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416]       |    12    |         11        |
| 5 | 416 |                      [1, 2, 4, 8, 16, 32, 64, 128, 256, 96, 192, 384, 416]                     |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416]       |    13    |         11        |
| 6 | 416 |                        [1, 2, 4, 8, 16, 32, 64, 128, 256, 192, 384, 416]                       |       [1, 2, 4, 8, 16, 32, 64, 128, 256, 384, 416]       |    12    |         11        |
| 2 | 675 |                  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 80, 336, 672, 3, 675]                 |  [1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675]  |    15    |         13        |
| 3 | 675 |                    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 640, 3, 643, 675]                   |  [1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675]  |    14    |         13        |
| 4 | 675 |                  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 3, 515, 80, 160, 675]                 |  [1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675]  |    15    |         13        |
| 5 | 675 |               [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 3, 160, 320, 640, 672, 675]              |  [1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675]  |    16    |         13        |
| 6 | 675 |               [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 320, 640, 17, 34, 35, 675]               |  [1, 2, 4, 8, 16, 32, 64, 128, 192, 224, 225, 450, 675]  |    16    |         13        |
| 2 | 694 |               [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 65, 321, 642, 646, 48, 694]              |  [1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694]  |    16    |         13        |
| 3 | 694 |                [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 640, 9, 18, 27, 54, 694]                |  [1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694]  |    16    |         13        |
| 4 | 694 |              [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 3, 6, 518, 80, 160, 176, 694]             |  [1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694]  |    17    |         13        |
| 5 | 694 |         [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 160, 320, 640, 672, 5, 10, 11, 22, 694]        |  [1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694]  |    19    |         13        |
| 6 | 694 |        [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 320, 640, 3, 6, 12, 13, 26, 27, 54, 694]        |  [1, 2, 4, 8, 16, 32, 64, 128, 144, 146, 274, 548, 694]  |    20    |         13        |
| 2 | 887 |               [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 68, 17, 273, 546, 819, 887]              | [1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887] |    16    |         14        |
| 3 | 887 |          [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 320, 832, 24, 48, 880, 3, 6, 7, 887]          | [1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887] |    19    |         14        |
| 4 | 887 |             [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 768, 17, 34, 51, 102, 119, 887]            | [1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887] |    17    |         14        |
| 5 | 887 |  [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 5, 10, 11, 22, 23, 96, 192, 384, 416, 832, 864, 887]  | [1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887] |    22    |         14        |
| 6 | 887 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 192, 384, 768, 832, 3, 6, 12, 13, 26, 27, 54, 55, 887] | [1, 2, 4, 8, 16, 32, 64, 96, 98, 99, 197, 394, 788, 887] |    23    |         14        |

#### Вывод

Таким образом, реализован алгоритм Яо, который может иметь цепочку длиной, равной или большей минимальной цепочки для заданного числа, в зависимости от числа и параметра k. Однако, алгоритм Яо отличается высокой скоростью работы, что является его преимуществом.


### Реализация алгоритма дробления вектора индексов

Реализуем алгоритм дробления вектора индексов с соответствием теоритическим положениям:

In [76]:
import time


def build_chain(vec) -> list :
    chain = [1]
    for i in vec:
        chain.append(chain[-1] + chain[i-1])
    return chain

def next_index_vector(vec, q=0) -> list:
    if vec == [1]*len(vec): return []
    for i in range(len(vec)-1, -1, -1):
        if vec[i] == 1:
            vec[i] = i + q + 1
        else:
            vec[i] -= 1
            break
    return vec

def dvi(n) -> list:
    
    if n == 1: return [1]
    if n <= 1: return []
    
    lower_border = ceil(log(n, 2))
    upper_border = floor(log(n, 2)) + bin(n).count('1')
    
    for m in range(lower_border, upper_border+1):
        q = m // 2 + 1
        index_vec = [*range(1, m+1)]
        
        mutable_ind = index_vec[:q]
        immutable_ind = index_vec[q:]
        
        while len(mutable_ind):
            chain = build_chain(mutable_ind+immutable_ind)
            
            if chain[-1] == n:
                return chain
            
            a_min = chain[q] + m - q
            a_max = chain[q] * (2^(m - q))
            
            if n > a_min and n < a_max :
                while len(immutable_ind):
                    chain = build_chain(mutable_ind+immutable_ind)
                    if chain[-1] == n:
                        return chain
                    immutable_ind = next_index_vector(immutable_ind, q)
                mutable_ind = next_index_vector(mutable_ind)
                immutable_ind = index_vec[q:]
            else:
                mutable_ind = next_index_vector(mutable_ind)
    return chain
                    


Проверим работу $n = 1152, 1028, 1344, 1005, 1008$, заодно замерив время работы:

In [77]:
import time
def test_dvi():
    n_list = [1152, 1028, 1344, 1005, 1008]
    res = []
    for n in n_list:
        start = time.time()
        chain = dvi(n)
        result_time = time.time() - start
        res.append((chain, result_time))
        print(f"n = {n} | {chain} | {round(result_time, 3)}s")
    return res
res = test_dvi() 

n = 1152 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152] | 0.001s
n = 1028 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028] | 0.001s
n = 1344 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344] | 1.273s
n = 1005 | [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 402, 804, 1005] | 428.316s
n = 1008 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008] | 8.075s


Результат в виде таблицы:
    
|   n  |                          Цепочка                          |          Время         |
|:----:|:---------------------------------------------------------:|:----------------------:|
| 1152 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1152]    |  0.001s |
| 1028 |    [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1028]    | 0.001s |
| 1344 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 1280, 1344] |    1.273s   |
| 1005 |       [1, 2, 4, 8, 16, 32, 64, 128, 192, 200, 201, 402, 804, 1005]       | 428.316s |
| 1008 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 320, 336, 672, 1008] | 8.075s
   |

#### Вывод
В первых трех примерах время расчетов не является репрезентативным, так как они специально подобраны для минимизации времени выполнения. В общем случае этот алгоритм более затратный по ресурсам (пример 4), чем предыдущая реализация алгоритма Яо, однако его основное преимущество состоит в том, что он гарантированно создает звездную цепочку наименьшей длины путем разделения вектора индексов. Таким образом, реализован алгоритм дробления вектора индексов. Как видно из таблицы, по сравнению с алгоритмом <em>Яо</em>, он является крайне ресурснозатратным, но зато гарантированно строит кратчайшую звёздную цепочку.

### Проверка гипотезы Шольца-Брауэра
Давайте проверим гипотезу Шольца-Брауэра для звездных цепочек с длинами, не превышающими 12, используя модифицированный алгоритм, названный "dvi_better". Исходя из теоретических положений, мы знаем, что для всех значений n, таких что $n \le 64$, выполняется равенство $l(2^{n}-1) = l(n) + n - 1$. Мы будем перебирать только те длины, которые равны $l(n) + n - 1$, чтобы определить, выполняется ли данная гипотеза и для наших значений.

In [59]:
def dvi_better(n,m) -> list:
    
    if n == 1: return [1]
    if n <= 1: return []
    
    m = m-1
    q = m//2+1
    
    index_vec = [*range(1, m + 1)]
    
    mutable_ind = index_vec[:q]
    immutable_ind = index_vec[q:]

    while len(mutable_ind):
        chain = build_chain(mutable_ind+immutable_ind)

        if chain[-1] == n:
            return chain

        a_min = chain[q] + m - q
        a_max = chain[q] * (2^(m - q))

        if n > a_min and n < a_max :
            while len(immutable_ind):
                chain = build_chain(mutable_ind+immutable_ind)
                if chain[-1] == n:
                    return chain
                immutable_ind = next_index_vector(immutable_ind, q)
            mutable_ind = next_index_vector(mutable_ind)
            immutable_ind = index_vec[q:]
        else:
            mutable_ind = next_index_vector(mutable_ind)
    return chain


Реализуем функцию для проверки данной гипотезы для $n \leq 12$:

In [84]:
def check_hypothesis():
    
    test_set = [*range(1, 13)]
    
    print("$n$|$l(2^n-1)$|$l(n) + n - 1$|$Result$")
    
    for elem in test_set:
        val_2 = len(dvi(elem)) + elem -1
        val_1 = len(dvi_better(2^elem - 1, val_2))
        if val_1 > val_2:
            print(f"{elem}|{val_1}|{val_2}|Не верно")
            break
        else:
            print(f"{elem}|{val_1}|{val_2}|Верно")

    
check_hypothesis()


$n$|$l(2^n-1)$|$l(n) + n - 1$|$Result$
1|1|1|Верно
2|3|3|Верно
3|5|5|Верно
4|6|6|Верно
5|8|8|Верно
6|9|9|Верно
7|11|11|Верно
8|11|11|Верно
9|13|13|Верно
10|14|14|Верно
11|16|16|Верно
12|16|16|Верно


Результат:

|   n |  l(2^n-1)  |  l(n) + n - 1  |  Result  |
|:---:|:----------:|:--------------:|:--------:|
|  1  |      1     |        2       |   Верно  |
|  2  |      3     |        3       |   Верно  |
|  3  |      5     |        5       |   Верно  |
|  4  |      6     |        6       |   Верно  |
|  5  |      8     |        8       |   Верно  |
|  6  |      9     |        9       |   Верно  |
|  7  |     11     |       11       |   Верно  |
|  8  |     11     |       11       |   Верно  |
|  9  |     13     |       13       |   Верно  |
|  10 |     14     |       14       |   Верно  |
|  11 |     16     |       16       |   Верно  |
|  12 |     16     |       16       |   Верно  |

#### Вывод

Итак, мы подтвердили, что гипотеза Шольца-Брауэра верна для цепочек с длинами не превышающими 12. Для ускорения проверки, мы воспользовались одним из следствий из теоретических положений, что позволило нам определить корректность гипотезы на основе значений сокращенных длин цепочек.

## Выводы
Мы получили представление о аддитивных цепочках и освоили навык составления и применения алгоритмов для нахождения минимальных аддитивных цепочек для заданного числа. Мы также освоили использование системы компьютерной математики SageMath для реализации этих алгоритмов. Конкретно, мы реализовали и проверили алгоритмы Яо и дробления вектора индексов, и провели сравнение эффективности этих алгоритмов.